### PR TITLE
Use backtrace formatting from the backtrace crate

### PR DESCRIPTION
**Pull request by cramertj**
r? @alexcrichton

### CODE REVIEW

**Comment by cramertj** [comment]
Probably want to block on https://github.com/rust-lang/backtrace-rs/pull/247 so we can clean this up before landing.

**Comment by alexcrichton** [comment]
I've published the next version with that PR, I'll take a closer look at this in an hour or so

**Comment by cramertj** [comment]
> Probably want to block on rust-lang/backtrace-rs#247 so we can clean this up before landing.

>I've published the next version with that PR, I'll take a closer look at this in an hour or so

I've updated this to use 1.37, thanks for the quick turnaround on that PR!

File: src/libstd/sys_common/backtrace.rs [hunk 1]
```diff
@@ -39,26 +37,74 @@ pub fn print(w: &mut dyn Write, format: PrintFormat) -> io::Result<()> {
...
+        output_filename(fmt, bows, print_fmt)
+    };
+    let mut bt_fmt = BacktraceFmt::new(fmt, print_fmt, &mut print_path);
+    bt_fmt.add_context()?;
+    let mut skipped = false;
     unsafe {
+        let mut idx = 0;
+        let mut res = Ok(());
         backtrace::trace_unsynchronized(|frame| {
+            if idx > MAX_NB_FRAMES {
```
Comment (cramertj) [inline src/libstd/sys_common/backtrace.rs]: Should this only happen if `PrintFmt::Short` is set? I didn't see that check in the original code, but it seems like it should be there, esp. given that if we truncate based on this, we also set `skipped` which causes the error messages saying to run with RUST_BACKTRACE=full (which the user may already be doing)

Comment (alexcrichton) [reply src/libstd/sys_common/backtrace.rs]: I think that this previously always happened but yeah let's only do it for short backtraces

File: src/libstd/sys_common/backtrace.rs [hunk 2]
```diff
@@ -39,26 +37,74 @@ pub fn print(w: &mut dyn Write, format: PrintFormat) -> io::Result<()> {
...
+                skipped = true;
+                return false;
+            }
+
+            // Remove the `17: 0x0 - <unknown>` line.
+            if print_fmt == PrintFmt::Short && frame.ip() == ptr::null_mut() {
```
Comment (alexcrichton) [inline src/libstd/sys_common/backtrace.rs]: This should be covered by the backtrace crate now

File: src/libstd/sys_common/backtrace.rs [hunk 3]
```diff
@@ -39,26 +37,74 @@ pub fn print(w: &mut dyn Write, format: PrintFormat) -> io::Result<()> {
...
+                skipped = true;
+                return true;
+            }
+
             let mut hit = false;
-            backtrace::resolve_frame_unsynchronized(frame, |symbol| {
-                hit = true;
-                printer.output(frame, Some(symbol));
-            });
+            if cfg!(not(target_os = "fuchsia")) { // Fuchsia can't symbolize in-proc
```
Comment (alexcrichton) [inline src/libstd/sys_common/backtrace.rs]: Is this clause necessary? While technically doing work that won't do much I think thisll work on fuchsia without the clause?

**Review by alexcrichton — COMMENTED** [review]
Looks great to me, thanks!

File: src/libstd/sys_common/backtrace.rs (hunk 2 above)
Comment (cramertj) [reply src/libstd/sys_common/backtrace.rs]: Oh, nice. Removed.

File: src/libstd/sys_common/backtrace.rs (hunk 3 above)
Comment (cramertj) [reply src/libstd/sys_common/backtrace.rs]: nope, not necessary, just saves some unnecessary work. Happy to take it out if you'd prefer.

File: src/libstd/sys_common/backtrace.rs (hunk 1 above)
Comment (cramertj) [reply src/libstd/sys_common/backtrace.rs]: Done.

File: src/libstd/sys_common/backtrace.rs (hunk 3 above)
Comment (alexcrichton) [reply src/libstd/sys_common/backtrace.rs]: Yeah if possible I think it's best to keep platform specific config away from here

Comment (cramertj) [reply src/libstd/sys_common/backtrace.rs]: Done.

File: src/libstd/panicking.rs [hunk 4]
```diff
@@ -167,9 +167,9 @@ fn default_hook(info: &PanicInfo<'_>) {
         let panics = update_panic_count(0);
 
         if panics >= 2 {
-            Some(backtrace::PrintFormat::Full)
+            Some(backtrace::PrintFmt::Full)
```
Comment (fbstj) [inline src/libstd/panicking.rs]:
```suggestion
            Some(backtrace_mod::PrintFmt::Full)
```
? or is `backtrace` the crates.io crate and `backtrace_mod` the 'rustc' crate?

Comment (cramertj) [reply src/libstd/panicking.rs]: This is coming from the `backtrace` crates.io crate. `std::backtrace` was imported here as `backtrace_mod` to distinguish between the two.

**Pull request merged**